File: hr_addon/hr_addon/report/work_hour_report/work_hour_report.py
# Copyright (c) 2022, Jide Olayinka and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
    columns, data = [], []
    condition_date, condition_employee = "", ""

    # Generate condition based on month and year filters
    if filters.get("month_filter") and filters.get("year_filter"):
        # Convert month name to month number
        month_number = {
            "January": 1, "February": 2, "March": 3, "April": 4,
            "May": 5, "June": 6, "July": 7, "August": 8,
            "September": 9, "October": 10, "November": 11, "December": 12
        }.get(filters.get("month_filter"))

        # Set the date range condition for the selected month and year
        year = filters.get("year_filter")
        condition_date = f"AND MONTH(log_date) = {month_number} AND YEAR(log_date) = {year}"

    if filters.get("employee_id"):
        empid = filters.get("employee_id")
        condition_employee += f" AND employee = '{empid}'"

    # Define columns for the report
    columns = [
        {'fieldname': 'log_date', 'label': 'Datum', 'width': 110},
        {'fieldname': 'name', 'label': 'Werktag (Link)', "fieldtype": "Link", "options": "Workday", 'width': 150},
        {'fieldname': 'total_work_seconds', 'label': 'Ist-Stunden', "width": 110},
        {'fieldname': 'total_target_seconds', 'label': 'Soll-Stunden', 'width': 110},
        {'fieldname': 'actual_diff_log', 'label': 'Differenz', 'width': 90},
    ]

    # Fetch data based on conditions
    work_data = frappe.db.sql(
        """
        SELECT name, log_date, employee, attendance, status, total_work_seconds, total_break_seconds,
               actual_working_hours * 60 * 60 actual_working_seconds,
               expected_break_hours * 60 * 60 expected_break_hours,
               target_hours, total_target_seconds, (total_work_seconds - total_target_seconds) AS diff_log,
               (actual_working_hours * 60 * 60 - total_target_seconds) AS actual_diff_log,
               TIME(first_checkin) AS first_in, TIME(last_checkout) AS last_out
        FROM `tabWorkday`
        WHERE docstatus < 2 {0} {1}
        ORDER BY log_date ASC
        """.format(condition_date, condition_employee), as_dict=1
    )

    data = work_data

    return columns, data

File: hr_addon/hr_addon/report/work_hour_report/work_hour_report.js
// Copyright (c) 2022, Jide Olayinka and contributors
// Copyright (c) 2024, Centura AG, Samuel Helbling, and contributors
// For license information, please see license.txt
// This file is licensed under the GNU Affero General Public License v3.0 (AGPL-3.0).
// See https://www.gnu.org/licenses/agpl-3.0.en.html for more details.
/* eslint-disable */
frappe.query_reports["Work Hour Report"] = {
    "filters": [
        {
            "fieldname": "month_filter",
            "label": __("Month"),
            "fieldtype": "Select",
            "options": [
                "January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"
            ],
			"default": new Date().toLocaleString('default', { month: 'long' }),
            "reqd": 1,
            "width": "35px"
        },
        {
            "fieldname": "year_filter",
            "label": __("Year"),
            "fieldtype": "Int",
            "default": parseInt((frappe.datetime.get_today()).split("-")[0]),
            "reqd": 1,
            "width": "35px"
        },
        {
            "fieldname": "employee_id",
            "label": __("Employee Id"),
            "fieldtype": "Link",
            "options": "Employee",
            "reqd": 1,
            "width": "35px"
        },
    ],
    "formatter": function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        // Using the new helper function for various time fields
        if (column.fieldname == "total_work_seconds" ||
            column.fieldname == "total_break_seconds" ||
            column.fieldname == "actual_working_seconds") {
            value = formatWithColor(value);
        }

        // Format Date with date_format defined in system-settings
        if (column.fieldname == "log_date") {
            if (value != 'Total') {
                let weekday = new Date(value).toLocaleString('en-US', { weekday: 'long' });
                weekday = __(weekday);

                value = frappe.datetime.str_to_user(value);
                value = `${weekday}, ${value}`;
            } else {
                value = parseInt((frappe.datetime.get_today()).split("-")[0]);
            }
        }

        // Special handling for diff_log and actual_diff_log with calDiff=true
        if (column.fieldname == "diff_log" || column.fieldname == "actual_diff_log") {
            value = formatWithColor(value, true);
        }

        // Other specific fields that just need the hitt function without color
        if (column.fieldname == "total_target_seconds" || column.fieldname == "expected_break_hours" || column.fieldname == "absent_seconds") {
            value = hitt(value);
        }

        return value;
    },
};

// Helper function to handle time formatting with colors
const formatWithColor = (value, calDiff = false) => {
    if (value == null) return ''; // Handle null/undefined values
    if (value < 0) {
        return "<span style='color:red'>" + hitt(value, calDiff) + "</span>";
    } else if (value > 0) {
        return "<span style='color:green'>" + hitt(value, calDiff) + "</span>";
    } else {
        return hitt(value, calDiff);
    }
};

const hitt = (fir, calDiff = false) => {
    if (fir < 0 && !calDiff) return fir; // Early return for negative and calDiff=false

    // Convert to positive if calDiff is true, otherwise keep as is
    let timeInSeconds = Math.abs(fir);

    let hours = Math.floor(timeInSeconds / 3600);
    let minutes = Math.floor((timeInSeconds % 3600) / 60);

    let result = `${hours > 0 ? hours + 'h ' : ''}${minutes > 0 ? minutes + 'm ' : ''}`.trim();

    // Return negative result if original fir was negative and calDiff is true
    return fir < 0 && calDiff ? `-${result}` : result;
};